ledger/eval/applications.go

```diff
@@ -292,18 +292,42 @@ func (cs *roundCowState) DelBox(appIdx basics.AppIndex, key string, appAddr basi
 	return true, cs.kvDel(fullKey)
 }
 
-func (cs *roundCowState) Perform(gi int, ep *logic.EvalParams) error {
+func (cs *roundCowState) Perform(gi int, ep *logic.EvalParams) (deltas *ledgercore.StateDelta, err error) {
+	cowForTxn := cs
+
+	if ep.GranularEval {
+		// In theory we could reuse this child cow allocation for all of the transactions in this
+		// group; that's what BlockEvaluator.TransactionGroup does. However, achieving the same
+		// thing here would require interface changes.
+
+		cowForTxn = &roundCowState{} // Specifically not calling cs.child(1), see comment below.
+		cs.reuseChild(cowForTxn, 1)
+		defer func() {
+			d := cowForTxn.deltas()
+			deltas = &d
+			cowForTxn.commitToParent()
+			// cowForTxn.recycle() // DO NOT RECYCLE!
+			// We cannot recycle here since we are returning the underlying StateDelta for this cow.
+			// If we did recycle, there would be a race between the caller looking at the StateDelta
+			// and the next user of the recycled cow. For this reason, we allocate a new
+			// roundCowState above instead of using cs.child(); otherwise, we would be draining the
+			// underlying sync.Pool and causing other routines to do more allocation. The other
+			// routines are likely more important, since if GranularEval is enabled, that's a good
+			// indicator that evaluation is for debugging purposes.
```

Comment:
idk maybe a little wordy here what about
```suggestion
			// cowForTxn.recycle() // DO NOT RECYCLE! as this is not a pool object, it is purposefully a fresh cow 
			// allocated on the stack to provide pool-free granular "per-transaction" deltas to tracers.
```